# Submission - Mismatch8

Mismatch kernel - n=8 - k=1 - Kernel2SVM - lbda = 10-4, 10-4, 10-5

In [1]:
import os
import sys
import numpy as np
import matplotlib.pyplot as plt

path = !pwd

base_dir = os.path.join(path[0], "../..")
sys.path.append(base_dir)

from utils.IOHandler import save_pickle, load_pickle

%matplotlib inline  

### Import data:

In [2]:
from utils.DataLoader import DataLoader

loader = DataLoader()
Xtr0, _, Ytr0, _ = loader.get_train_val(k=0, val_size=0.0, random_state=42)
Xtr1, _, Ytr1, _ = loader.get_train_val(k=1, val_size=0.0, random_state=42)
Xtr2, _, Ytr2, _ = loader.get_train_val(k=2, val_size=0.0, random_state=42)

### Compute kernel:

In [3]:
from src.kernels.MismatchKernel import MismatchKernel

#mismatch_kernel = MismatchKernel(n=8, k=1, charset="ATCG", verbose=0)
#save_pickle("mismatch_kernel_n7", mismatch_kernel)

mismatch_kernel = load_pickle("mismatch_kernel_n8")

#gram_tr_tr_0 = mismatch_kernel(Xtr0, Xtr0)
#gram_tr_tr_1 = mismatch_kernel(Xtr1, Xtr1)
#gram_tr_tr_2 = mismatch_kernel(Xtr2, Xtr2)


In [4]:
#np.save('gram_tr_tr_0_cross_val', gram_tr_tr_0)
#np.save('gram_tr_tr_1_cross_val', gram_tr_tr_1)
#np.save('gram_tr_tr_2_cross_val', gram_tr_tr_2)
gram_tr_tr_0 = np.load('gram_tr_tr_0_cross_val.npy')
gram_tr_tr_1 = np.load('gram_tr_tr_1_cross_val.npy')
gram_tr_tr_2 = np.load('gram_tr_tr_2_cross_val.npy')

### Perform classification:

#### Kernel 2SVM:

In [5]:
from src.classifiers.Kernel2SVM import Kernel2SVM
from src.evaluation.metrics import *

#### Cross-validation:

#### Dataset 0:

In [6]:
from src.evaluation.selection import cross_validate

In [7]:
list_lbdas = np.logspace(-4, -2, 9)
for lbda in list_lbdas:
    print("\nLambda = {}".format(lbda))
    kernel_2svm = Kernel2SVM(lbda=lbda, support_vec_tol=1e-2, verbose=0)
    _ = cross_validate(kernel_2svm, gram_tr_tr_0, Ytr0, cv=5, scoring="accuracy_score")


Lambda = 0.0001
Mean accuracy = 0.609 - Std = 0.007516648189186478

Lambda = 0.00017782794100389227
Mean accuracy = 0.6140000000000001 - Std = 0.013285330255586422

Lambda = 0.00031622776601683794
Mean accuracy = 0.624 - Std = 0.020161845153655957

Lambda = 0.0005623413251903491
Mean accuracy = 0.6295 - Std = 0.012389511693363848

Lambda = 0.001
Mean accuracy = 0.626 - Std = 0.0037416573867739005

Lambda = 0.0017782794100389228
Mean accuracy = 0.6134999999999999 - Std = 0.013656500283747637

Lambda = 0.0031622776601683794
Mean accuracy = 0.626 - Std = 0.020099751242241757

Lambda = 0.005623413251903491
Mean accuracy = 0.609 - Std = 0.018881207588499215

Lambda = 0.01
Mean accuracy = 0.6165 - Std = 0.013928388277184081


$\lambda = 0.00031622776601683794$ - Accuracy: 0.627

#### Dataset 1:

In [8]:
list_lbdas = np.logspace(-5, -2, 9)
for lbda in list_lbdas:
    print("\nLambda = {}".format(lbda))
    kernel_2svm = Kernel2SVM(lbda=lbda, support_vec_tol=1e-2, verbose=0)
    _ = cross_validate(kernel_2svm, gram_tr_tr_1, Ytr1, cv=5, scoring="accuracy_score")


Lambda = 1e-05
Mean accuracy = 0.7474999999999999 - Std = 0.007745966692414834

Lambda = 2.3713737056616554e-05
Mean accuracy = 0.7384999999999999 - Std = 0.013190905958272891

Lambda = 5.623413251903491e-05
Mean accuracy = 0.7365 - Std = 0.009823441352194269

Lambda = 0.0001333521432163324
Mean accuracy = 0.7450000000000001 - Std = 0.02043281674170255

Lambda = 0.00031622776601683794
Mean accuracy = 0.7615000000000001 - Std = 0.021482551058940827

Lambda = 0.0007498942093324559
Mean accuracy = 0.7420000000000001 - Std = 0.011113055385446425

Lambda = 0.0017782794100389228
Mean accuracy = 0.7430000000000001 - Std = 0.021988633427296022

Lambda = 0.004216965034285823
Mean accuracy = 0.7304999999999999 - Std = 0.017986105748604948

Lambda = 0.01
Mean accuracy = 0.7275 - Std = 0.01949358868961795


$\lambda = 0.0007498942093324559$ - Accuracy: 0.7484999999999999

#### Dataset 2:

In [9]:
list_lbdas = np.logspace(-6, -3, 9)
for lbda in list_lbdas:
    print("\nLambda = {}".format(lbda))
    kernel_2svm = Kernel2SVM(lbda=lbda, support_vec_tol=1e-2, verbose=0)
    _ = cross_validate(kernel_2svm, gram_tr_tr_2, Ytr2, cv=5, scoring="accuracy_score")


Lambda = 1e-06
Mean accuracy = 0.6464999999999999 - Std = 0.021598611066455158

Lambda = 2.3713737056616552e-06
Mean accuracy = 0.654 - Std = 0.012103718436910195

Lambda = 5.623413251903491e-06
Mean accuracy = 0.6475000000000001 - Std = 0.022527760652137633

Lambda = 1.333521432163324e-05
Mean accuracy = 0.641 - Std = 0.024269322199023193

Lambda = 3.1622776601683795e-05
Mean accuracy = 0.6495000000000001 - Std = 0.008426149773176357

Lambda = 7.498942093324559e-05
Mean accuracy = 0.655 - Std = 0.013322912594474235

Lambda = 0.00017782794100389227
Mean accuracy = 0.6605000000000001 - Std = 0.023313086453749524

Lambda = 0.00042169650342858224
Mean accuracy = 0.657 - Std = 0.02135415650406264

Lambda = 0.001
Mean accuracy = 0.65 - Std = 0.00836660026534077


$ \lambda = 1.333521432163324e-05$ - Accuracy: 0.6535

| Dataset | 1 | 2 | 3 |
|---||---|---|---|
|Accuracy| $0.627$ | $0.748$  | $0.653$  |
|$\lambda$| $3.16 10^{-4}$ | $4.50 10^{-4}$  | $1.33 10^{-4}$  |

#### Training and prediction for submission:

In [ ]:
kernel = load_pickle("mismatch_kernel_n8")
dataloader = DataLoader()
list_preds = []

for k in range(3):
    print("Loading dataset number {} and training the model...".format(k))
    model = Kernel2SVM(lbda=lbdas[k], support_vec_tol=1e-2, verbose=0)
    x_train, _, y_train, _ = dataloader.get_train_val(k, 0., 42)
    x_test = dataloader.get_test(k)
    gram_train = kernel(x_train, x_train)
    model.fit(gram_train, y_train)
    print("Inference of the model...".format(k))
    gram_test = kernel(x_train, x_test)
    y_pred_test = model.predict(gram_test)
    y_pred_test[y_pred_test == -1] = 0
    y_pred_test = y_pred_test.astype(int)
    
    list_preds += y_pred_test.tolist()

    with open("submission_mismatch_8.csv", 'w') as f:
        f.write('Id,Bound\n')
        for i in range(len(list_preds)):
            f.write(str(i)+','+str(list_preds[i])+'\n')